### PR TITLE
Add redux to local storage

### PR DESCRIPTION
**Pull request by Tylyps**
Add saveing all changes on tasks to localstorage in redux

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -8,6 +8,9 @@ import App from './App';
 import registerServiceWorker from './registerServiceWorker';
 
 const store = storeConfig();
+store.subscribe(() => {
+    localStorage.setItem('reduxState', JSON.stringify(store.getState()))
```
Comment (Lukasz-Semik) [inline src/index.js]: Currently, localStorage things are not managed by action and reducer.
the idea of this task is to use middleware. In example you can use `redux-thunk`.

This is very dangerous approach in terms of `localStorage`. 
`localStorage` exists only in window env, and it is not supported by 100% browser (please, read about it and please provide me info, what is the `%` of usage currently).
It means, that this app will be totally broken on not supporting browsers (due to `undefined`).
Please, move this to the action (hint-> middleware) and also protect `localStorage` itself.

File: src/store/storeConfig.js [hunk 2]
```diff
@@ -2,11 +2,15 @@ import { createStore, combineReducers } from 'redux';
...
 
+const persistedState = localStorage.getItem('reduxState') ? JSON.parse(localStorage.getItem('reduxState')) : {}
+
 const storeConfig = () => {
     const store = createStore(combineReducers({
         exampleReducer,
         list,
-    }));
+    }),
+    persistedState,
```
Comment (Lukasz-Semik) [inline src/store/storeConfig.js]: it seems that format of line 11, 12, 13 together are broken.

File: src/store/storeConfig.js [hunk 3]
```diff
@@ -2,11 +2,15 @@ import { createStore, combineReducers } from 'redux';
 import exampleReducer from './reducers/exampleReducer';
 import list from './reducers/list';
 
+const persistedState = localStorage.getItem('reduxState') ? JSON.parse(localStorage.getItem('reduxState')) : {}
```
Comment (Lukasz-Semik) [inline src/store/storeConfig.js]: this is not redux approach. Got data from `localStorage` are those one, which should go via reducers and passed to state by `combineReducers`.

**Review by Lukasz-Semik — COMMENTED** [review]
left some comments.
I guess that you meant `Add localStorage to redux`, hmm ? :P

File: src/index.js (hunk 1 above)
Comment (Lukasz-Semik) [reply src/index.js]: To make this comment full:
- lack of localStorage protection will broke tests
- lack of localStorage will broke server side rendering
as in test env, and on server, there is no window at all.

Comment (Tylyps) [reply src/index.js]: The % of usage localStorage by browser is 93.75%, for today almost all browser support Web Storage (sessionStorage and localStorage) only opera mini and old versions of opera (from 2009), safari (2008) and IE(2006).

**Review by Lukasz-Semik — CHANGES_REQUESTED** [review]
1. Browser support is not the only reason to protect the localStorage. If app would be server side rendered, will be broken. Also, tests will be broken. Even if we have super high usage here, we should be safe and not totally break (blank page) the app on old browsers. This is what `making good apps` means.

2. Second point - this is not what this task is about. This task was created to understand and learn how to use middlewares. So:
- please remove `redux-localstorage...`
- learn about middleware (it is needed to make a proper approach with fetching from api). 
- items are not deleted from localStorage
- use redux-thunk and move prepared functions touching localStorage to middlewares. It should return proper actions generators.

File: src/store/actions/listActions.js [hunk 4]
```diff
@@ -1,15 +1,37 @@
...
     type: REMOVE_TASK,
     payload: {
         id,
     }
 });
+
+export const removeTask = id => (
+    dispach => {
+        dispach(removeTaskAsync(id))
+        setTimeout(() => {
```
Comment (Lukasz-Semik) [inline src/store/actions/listActions.js]: do we need this `setTimeout` here?

File: src/store/reducers/list.js [hunk 5]
```diff
@@ -13,6 +13,9 @@ const list = (state = initialState, action) => {
                 ...state,
                 tasks: state.tasks.filter((task, id) => id !== action.payload.id ),
             }
+        case SAVE:
+            localStorage.setItem('list',JSON.stringify(state))
```
Comment (Lukasz-Semik) [inline src/store/reducers/list.js]: it should be in actions file, inside of the the function, which returns the action generator function

File: src/store/reducers/list.js [hunk 6]
```diff
@@ -1,5 +1,5 @@
-import {ADD_TASK, REMOVE_TASK} from '../variables';
-const initialState = { tasks: ['First task'] };
+import {ADD_TASK, REMOVE_TASK, SAVE} from '../variables';
+const initialState = JSON.parse(localStorage.getItem('list')) || { tasks: [] };
```
Comment (Lukasz-Semik) [inline src/store/reducers/list.js]: initial state should `{ tasks: []}`, taking from localStorage should be handled via `componentDidMount`

File: src/store/variables.js [hunk 7]
```diff
@@ -3,3 +3,5 @@ export const MULTIPLY_ACTION = 'MULTIPLY_ACTION';
 
 export const ADD_TASK = 'ADD_TASK';
 export const REMOVE_TASK = 'REMOVE_TASK';
+
+export const SAVE = 'SAVE';
```
Comment (Lukasz-Semik) [inline src/store/variables.js]: I gess, that we are saving the task.

**Review by Lukasz-Semik — COMMENTED** [review]
- `localStorage` is still not protected.
- I don't understand the idea in this saving. It should be done in simple way:
Adding: when adding the task, at first, save in localStorage and then dispatch the adding task section.
Removing: analogical to Adding.
Getting all: another action + function returning action generator, which is checking if there are some items in localStorage and return if they exists

File: src/store/actions/listActions.js [hunk 8]
```diff
@@ -7,9 +24,30 @@ export const addTask = task => ({
...
     type: REMOVE_TASK,
     payload: {
         id,
     }
 });
+
+export const startRemoveTask = id => (
+    (dispach, getState) => {
+        const tasks = getState().list.tasks.filter((task, taskId) => taskId !== id );
+        if(localStorage) {
```
Comment (Lukasz-Semik) [inline src/store/actions/listActions.js]: this piece of code seems to be redundant. That's why I requested special helper -> it should be created a file with helpers or utils, where all localStorage relations are managed. Then, only import here this function and use.

File: src/store/storeConfig.js [hunk 9]
```diff
@@ -1,12 +1,16 @@
...
-    const store = createStore(combineReducers({
-        exampleReducer,
-        list,
-    }));
+    const store = createStore(
+        combineReducers({
+            exampleReducer,
+            list,
+        }),
+        applyMiddleware(thunk)
```
Comment (Lukasz-Semik) [inline src/store/storeConfig.js]: I am wondering, how redux devtools are set up in this project.

File: src/helper/localstorage.js [hunk 10]
```diff
@@ -0,0 +1,13 @@
+export const saveToLocalStorage = (key, value) => {
+    if(localStorage) {
+      localStorage.setItem(key, JSON.stringify(value))
+  }
+}
+
+export const loadFromLoaclStorage = (key) => {
```
Comment (Lukasz-Semik) [inline src/helper/localstorage.js]: typo

**Review by Lukasz-Semik — APPROVED** [review]

**Pull request merged**